:bug: [#1653] fix OrderingFilter query

Floris272 opened the pull request:
Closes #1653

**Changes**

- order_by & distinct set by the viewset queryset was overridden in the OrderFilter

**Checklist**

Check off the items that are completed or not relevant.

- Experimental features/changes

  - [x] Any experimental features added in this PR are backwards compatible
  - [x] Any experimental features added in this PR are documented in the `docs/api/experimental.rst` page

- Commit hygiene

  - [x] Commit messages refer to the relevant Github issue
  - [x] Commit messages explain the "why" of change, not the how

Floris272 committed:
:bug: [#1653] fix OrderingFilter query

stevenbal reviewed (changes requested):
@Floris272 could you add a unittest for this?

Floris272 committed:
:white_check_mark: [#1653] add ordering filter test for dupicate versions

stevenbal reviewed (changes requested):
While this does fix the issue, the query seems to be very inefficient now (locally I have 1 million ObjectInformatieObjecten in the db, not sure how it performs for lower numbers)

before:

<img width="960" height="646" alt="image" src="https://github.com/user-attachments/assets/c813fff2-480b-4d8c-8d73-a925450ab57b" />

after:

<img width="960" height="854" alt="image" src="https://github.com/user-attachments/assets/6d7d9300-b041-4e88-bdee-4d1a56aef7b5" />

Could you add a basic performance test for this scenario in `performance_test` (let's make a separate file for it) and make sure the main queries are more performant?

Floris272 committed:
:zap: [#1653] add original qs select & prefetch related args to new query

Floris272 committed:
:white_check_mark: [#1653] add eio list and with ordering filter performance test

Floris272 committed:
:green_heart: [#1653] fix eio performance test urls

stevenbal reviewed (changes requested):
@Floris272 unfortunately the performance does seem to be quite dependent on the number of EIOs, I tested locally with 500k on main and on your branch, and the queries still seem quite slow :(

<img width="960" height="602" alt="image" src="https://github.com/user-attachments/assets/a5af508f-3eda-42a4-8247-27bce4777653" />


and after the changes

<img width="960" height="629" alt="image" src="https://github.com/user-attachments/assets/a2e7b5e1-dca7-4d93-9a7f-dc617bc7b6e8" />

Could you check if it's possible to fix this?

Floris272 committed:
:recycle: [#1653] move dupicate code to conftest.py

Floris272 committed:
:recycle: [#1653] fix imports

Floris272 commented:
@stevenbal haven't been able to find a better query for this. The reason the subquery is needed because a new order_by will override the one on the viewset which is needed for the distinct call.

CharString commented on src/openzaak/utils/filterset.py:
```diff
@@ -15,7 +15,12 @@ def filter(self, qs, value):
 
         ordering = [self.get_ordering_value(param) for param in value]
-        return qs.order_by(*ordering).distinct()
+        return (
```
@stevenbal Yes. With the current data model, there is no way around the full scan. So we can either 


- pre-compute and update a materialised view of the 
  ```sql
  SELECT DISTINCT ON(canonical_id) *
  FROM documenten_enkelvoudiginformatieobject
  ORDER BY canonical_id, versie DESC
  ```
  and inner join that view. We'd need to periodically update.
- we change `EnkelvoudigInformatieObjectCanonical.latest_version` from a property to a field, and do an upsert to `GREATEST(latest_version, %s)`, eio.version when we save EIOs.
- we could disallow sorting without a filter. Because as you noted, the full scan is ok-ish if it's just over a small subset.


I looked into the first options; they have different trade-offs. But I discovered CMIS tests were skipped locally. So I don't have a full view on the effect on that.

Floris272 committed:
:recycle: [#1653] add latest_version field to EnkelvoudigInformatieObjectCanonical

Floris272 committed:
:recycle: [#1653] add post_save to signal

Floris272 committed:
:recycle: [#1653] call post_save in generate_data command

Floris272 committed:
:recycle: [#1653] add documenten-only flag to generate_data

stevenbal commented on src/openzaak/components/documenten/api/viewsets.py:
```diff
@@ -314,6 +309,14 @@ def filterset_class(self):
+        if (
+            self.detail
+        ):  # detail needs to be able to filter by versie for VERSIE_QUERY_PARAM
+            return queryset.order_by("canonical", "-versie").distinct("canonical")
```
Not sure if you tried this, but I *think* getting the PKs of the latest versions and then filtering by that might be more efficient, because `DISTINCT` can be a costly operation:

```suggestion
            return queryset.filter(id__in=EnkelvoudigInformatieObjectCanonical.objects.values("latest_version_id"))
```

stevenbal commented on src/openzaak/components/documenten/models.py:
```diff
@@ -269,15 +269,17 @@ class EnkelvoudigInformatieObjectCanonical(models.Model):
         help_text="Hash string, wordt gebruikt als ID voor de lock",
     )
 
+    latest_version = models.OneToOneField(
```
We're going to need a data migration to populate this for existing instances. I think we can do it in a single query, but it would be good to check how this performs for bigger datasets, because long running datamigrations have been a problem before in Objects API

Floris272 committed:
:recycle: [#1653] add latest_version migration code & change signal to post_delete

Floris272 committed:
:recycle: [#1653] change generate_data documents_only to resources flag

Floris272 committed:
:white_check_mark: [#1653] refresh canonical

Floris272 committed:
:recycle: [#1653] change latest_version on_delete to SET_NULL

CharString reviewed (commented):
This approach brings the performance of a sorting query back in the same ballpark:

<img width="484" height="852" alt="Image" src="https://github.com/user-attachments/assets/0172eea6-8e27-4c2a-9b7a-9306bfa3633e" />

But this pagination query:

<img width="1431" height="195" alt="Image" src="https://github.com/user-attachments/assets/af4f09a1-920a-42e2-9bea-a155c8398279" />

Is a very long winded, slow version of:
```sql
SELECT COUNT(*) FROM documenten_enkelvoudiginformatieobjectcanonical;
```

And joining twice here:

<img width="1348" height="183" alt="Image" src="https://github.com/user-attachments/assets/c88f1002-c703-423d-a016-e16d1d1bb589" />
still leaves a bit of performance on the table. (T4 is exactly the same as the first JOIN to the same table, but the planner can't know that)

But changing the viewset to use the canonical as the basis isn't as trivial as I thought, with that whole stack of mixins. (and the text in the OAS is cut-off right where it is about to explain how multiple versions are to be queried, so the spec is unclear to me)

I did check the query plans for cleaned up queries. The count is laughably faster (±35 ms and the difference will be bigger if there are many eio's with multiple versions), as it only hits the primary index. The de-duplicated join is about as fast as the quicker (but wrong) query in main.

stevenbal commented on src/openzaak/management/commands/generate_data.py:
```diff
@@ -7,6 +7,7 @@
 from django.contrib.gis.geos import Point
 from django.core.management.base import BaseCommand, CommandError
 from django.db import models, transaction
+from django.db.models.signals import post_save
```
Not sure why this happened, but I don't run into it anymore after regenerating data

Floris272 committed:
:memo: [#1653] fix cutoff api spec description

Floris272 commented:
@CharString fixed the api spec comment (specific versions can be fetched using query string params) and changed `DO_NOTHING` to `SET_NULL` as it will indeed crash if the signal on_delete does not fire.

CharString reviewed (changes requested):
> @CharString fixed the api spec comment (specific versions can be fetched using query string params) and changed `DO_NOTHING` to `SET_NULL` as it will indeed crash if the signal on_delete does not fire.

Okay. Crashing is bad... But rolling back the transaction was correct. Setting it to NULL is exactly the wrong behaviour @stevenbal was worried about: iff the signal doesn't fire correctly, the key reference is incorrect.

The tightest thing to do is do it in Postgres as an AFTER DELETE TRIGGER.
The easiest thing to do is let it crash and live on with correct data, but client app may not retry after a 500.
The most familiar thing to do that could possibly work, is to keep the SET_NULL and schedule (on crash/after delete/periodically) a variant of the "set latest version" query that runs on the NULL rows.

Floris272 committed:
:sparkles: [#1653] add tast to periodically update canonical latest_version if incorrect

Floris272 committed:
:recycle: [#1653] replace on_delete signal with db trigger

Floris272 committed:
:white_check_mark: [#1653] fix EnkelvoudigInformatieObjectCanonicalFactory

Floris272 committed:
:recycle: [#1653] add canonical refresh to eio delete

CharString committed:
:recycle: refactor: Move refresh from delete to __str__

API DELETE requests return 204 no content, so they won't have stale
data.

Therefore making every delete extra expensive with a refresh is a bit
unneeded for the general case.

The admin *does* bump into stale cache when displaying after a delete,
so we put a just in time refresh in place.

CharString commented:
@Floris272  [d5f65d3](https://github.com/open-zaak/open-zaak/commit/d5f65d3760fa8c49ce97dd62b4800f4c9add9f65)  let's ship it. :)
